src/main/java/LinkedBinarySearchTree.java

```diff
@@ -0,0 +1,492 @@
+        if (child == null) {
+            throw new NoSuchElementException();
+        } else if (child.getData().equals(element)) {
+            if (parent.getData().compareTo(element) > 0) {
+                parent.setLeft(findReplacementNode(child));
+            } else {
+                parent.setRight(findReplacementNode(child));
+            }
+            return child.getData();
+        } else if (child.getData().compareTo(element) > 0) {
+            return remove(element, child, child.getLeft());
+        } else {
+            return remove(element, child, child.getRight());
+        }
+    }
+
+    /**
+     * Helper method for finding which node should be used to replace a node being removed.
+     * There are a few conditions:
+     * <p>
+     * 1. Both left and right children are null. Here, simply remove the node.
+     * <p>
+     * 2. Left child is not null but right child is null. Replace the toRemove node with the left child.
+     * <p>
+     * 3. Left child is null but right child is not null. Replace the toRemove node with the right child
+     * <p>
+     * 4. Both left and right children are not null. Replace the toRemove node with the in order successor,
+     * which would be the right subtree's left most node.
+     *
+     * @param toRemove The node being replaced
+     * @return The node replacing the node being removed
+     */
+    private Node<T> findReplacementNode(Node<T> toRemove) {
+        Node<T> replacementNode = null;
+        if (toRemove.getLeft() == null && toRemove.getRight() == null) {
+            replacementNode = null;
+        } else if (toRemove.getLeft() != null && toRemove.getRight() == null) {
+            replacementNode = toRemove.getLeft();
+        } else if (toRemove.getLeft() == null && toRemove.getRight() != null) {
+            replacementNode = toRemove.getRight();
+        } else {
+            Node<T> parent = toRemove;
+            Node<T> child = toRemove.getRight();
+            // Find the in order successor (right child's left
+            // most node (minimum node))
+            while (child.getLeft() != null) {
+                parent = child;
+                child = child.getLeft();
+            }
+            // Set replacement node's left to
+            // the node being removed's (subtree root's) left
+            child.setLeft(toRemove.getLeft());
+            // If the immediate in order successor is NOT the
+            // node being replaced's right child, the parent
+            // node's new left becomes the child node's right
+            // and the child node's right is replaced with
+            // the node being replaced's right
+            if (toRemove.getRight() != child) {
+                parent.setLeft(child.getRight());
+                child.setRight(toRemove.getRight());
+            }
+            replacementNode = child;
+        }
+        return replacementNode;
+    }
+
+    @Override
+    public T removeMin() {
```

Comment:
These methods can be simplified quite a bit if you also maintain `node.parent`.
* write `Node<T> minNode()` (which can be used for `min()`, `minNode().data` (logic sharing!)
* `removeMin` becomes `minNode().parent.left = null` plus storage for return value

`add` doesn't become much more complex either, only the base case needs to be changed to set the `parent`. `remove` is probably the worst, but also only the updates for `toRemove.children.parent = toRemove.parent` or however you wish to balance.

Reply:
This is absolutely right, but I think I will leave this out of the main bit on the BSTs. 

As of now I am thinking that I may do an aside introducing the idea and then have an assignment or lab have them implement it.

Reply:
Fair enough, sounds good.